\documentclass{article}
\usepackage{hyperref}
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{subcaption}
\usepackage[symbol]{footmisc}
\usepackage{todonotes}
\bibliographystyle{nature}


%\title{libbdsg and libhandlegraph: High-performance sequence graph implementations for graphical pangenomics}
\title{} % for graphical pangenomics}
  %libbdsg and libhandlegraph: High-performance sequence graph implementations for graphical pangenomics}
\author{}

\begin{document}

\maketitle

\begin{abstract}


\end{abstract}

\section{Introduction}


few samples does not allow association 



\section{Results}

\subsection{Pipeline} 


Per sample processing 
Combine stats that are not correlated (supplementary)

Genic  

regulatory 


Variants need to be in genic regions, have allele freqency<0.05\% in the reference populations, impact low or greater and either be in genes belonging to at least two lists of relevant genes or have CADD socore above the 90th percentile and the gene must have pLI>90. 

[1] "f4\tTot Number of Uniqe Variants\t373"
[1] "f4\tTot Number of NOB\t38"
[1] "f4\tAverage number of sites per sample\t70.8333333333333"
[1] "f4\tSd number of sites per sample\t22.9034204141361"
[1] "f4\tTotal number of Unique Genes\t253"
[1] "f4\tTotal number of Unique Transcripts\t253"
[1] "f4\tAverage number of genes per sample\t55.6666666666667"
[1] "f4\tSd number of genes per sample\t17.3627954738477"

Filter retains 373 uniqe variants of which four with high impact located in 253 unique transcripts correspinding to 253 genes. Corresponding samples average are 70.8 variants (s.d.=22.9) in 55.6 genes (s.d.=17.36).  


with high impact  in YY (s.d. XX) genes per samples (zz unique genes). Type of consequences shown in B 



\section{Discussion}


\section{Methods}